### PR TITLE
Show traceids for failing and successful requests

### LINKED ISSUE #64245: provide the current backend-traceID to the frontend
it would be very useful if the frontend (in browser code) would receive the current traceId in the http-response of a datasource request:
- we could show the current traceId in error-messages, allowing the users to easily look up the trace for the error
- we could also provide this info for non-error-messages (perhaps somewhere in the query inspector)

questions:
- is it always one or zero traceIds for 1 ajax-request? can somehow 2 or more traceIds "happen"?

### PR DESCRIPTION
**Pull request by gabor**
( fixes https://github.com/grafana/grafana/issues/64245 )

the grafana-go-server is able to do tracing, when configured so. but, it's not possible to discover the current `traceId` in the frontend (and then for example communicate it to support)

two scenarios:
- a query fails, the user should be able to discover the traceId from the error message.
- a query succeeds, but the user still needs a reference to it, perhaps it was too slow

the plan with this PR is to allow the user to see the current traceId, in the following ways:
- for failed queries
    - in explore
        - the traceId is shown as part of the error-message
        - the traceId is also shown in the query-inspector/errors tab
    - in dashboard
       - when building the query, the traceId is shown as part of the error-message
       - the traceId is also shown in the query-inspector/errors tab
       - if the user clicks on the red-triange-in-the-upper-left-corner-of-a-panel, they get sent to the query-inspector/errors tab (where they can see the traceId)
- for successful queries
   - in query-inspector/stats tab, at the bottom (technically, both failed and successful query traceIds appear here)

the way the PR works is:
- in the `go` code we add the traceId to the ajax-response-headers
- in the frontend, we read out this header, and then push it through every necessary layer to be able to visualize it
- the visualization itself, is behind a feature flag, named `showTraceId` (note, the rest, like sending the traceId from the backend, etc... is not behind a feature flag)

i'm open to any suggestion about how to achieve this in a better way, so feel free to comment 👍 .

(i'll add more unit tests when i see the reviewers agree with the approach, i just did not want to spend the effort on them when i may have to completely rework this thing, sorry)

### CODE REVIEW

**Comment by bboreham** [comment]
> the grafana go server adds a special HTTP header to the ajax-responses with the traceId

Suggest using W3C-style `traceparent`.  https://www.w3.org/TR/trace-context/#header-name

**Comment by gabor** [comment]
> > the grafana go server adds a special HTTP header to the ajax-responses with the traceId
> 
> Suggest using W3C-style `traceparent`. https://www.w3.org/TR/trace-context/#header-name

thanks for the idea, applied.

File: packages/grafana-data/src/types/datasource.ts [hunk 1]
```diff
@@ -465,6 +465,11 @@ export interface DataQueryResponse {
    * Defaults to LoadingState.Done if state is not defined
    */
   state?: LoadingState;
+
+  /**
+   * traceIds related to the response, if available
+   */
+  traceIds?: string[];
```
Comment (marefr) [inline packages/grafana-data/src/types/datasource.ts:472]: This could only be a single one at most, right?

**Comment by svennergr** [comment]
> > > the grafana go server adds a special HTTP header to the ajax-responses with the traceId
> > 
> > 
> > Suggest using W3C-style `traceparent`. https://www.w3.org/TR/trace-context/#header-name
> 
> thanks for the idea, applied.

@gabor / @bboreham I am not sure if that header name really makes sense here, also see https://www.w3.org/TR/trace-context/#other-risks
With this PR we are not really starting a trace in the frontend yet, but would always return that header from Grafana.

File: public/app/features/inspector/InspectErrorTab.tsx [hunk 2]
```diff
@@ -36,6 +36,12 @@ function renderError(error: DataQueryError) {
         <>
           {error.status && <>Status: {error.status}. Message: </>}
           {msg}
+          {error.traceId != null && (
+            <>
+              <br />
+              (traceId: {error.traceId})
```
Comment (svennergr) [inline public/app/features/inspector/InspectErrorTab.tsx:42]:
```suggestion
              (Trace ID: {error.traceId})
```

File: public/app/features/inspector/InspectStatsTab.tsx [hunk 3]
```diff
@@ -59,11 +60,14 @@ export const InspectStatsTab = ({ data, timeZone }: InspectStatsTabProps) => {
 
   const statsTableName = t('dashboard.inspect-stats.table-title', 'Stats');
   const dataStatsTableName = t('dashboard.inspect-stats.data-title', 'Data source stats');
+  const traceIdsStatsTableName = t('dashboard.inspect-stats.data-traceids', 'Trace ids');
```
Comment (svennergr) [inline public/app/features/inspector/InspectStatsTab.tsx:63]: Not sure about this.

```suggestion
  const traceIdsStatsTableName = t('dashboard.inspect-stats.data-traceids', 'Trace IDs');
```

File: public/app/features/query/components/QueryErrorAlert.tsx [hunk 4]
```diff
@@ -18,7 +18,10 @@ export function QueryErrorAlert({ error }: Props) {
       <div className={styles.icon}>
         <Icon name="exclamation-triangle" />
       </div>
-      <div className={styles.message}>{message}</div>
+      <div className={styles.message}>
+        {message} <br />
+        (traceId: {error.traceId})
```
Comment (svennergr) [inline public/app/features/query/components/QueryErrorAlert.tsx:23]:
```suggestion
        (Trace ID: {error.traceId})
```

**Comment by bboreham** [comment]
> @gabor / @bboreham I am not sure if [traceparent] really makes sense here, also see https://www.w3.org/TR/trace-context/#other-risks
> With this PR we are not really starting a trace in the frontend yet, but would always return that header from Grafana.

The frontend is still participating, though it didn't originate the trace.
Even when the frontend starts to generate spans, not all traces will originate in the frontend, e.g. if you open a Grafana URL from Slack.

It would strike me as very weird to use a header but deliberately pick a non-standard name.

File: packages/grafana-data/src/types/datasource.ts (hunk 1 above)
Comment (marefr) [reply packages/grafana-data/src/types/datasource.ts:472]: Understood

**Comment by sakjur** [comment]
> It would strike me as very weird to use a header but deliberately pick a non-standard name.

`traceparent` seem to have a rather strict formatting requirement that we should make sure to follow if we intend to use it. If the alternative is between incorrectly implementing parts of a standard and deliberately picking a non-standardized name, I'd prefer the latter. Even better in my mind is complying with the standard unless there is a strong and valid reason not to.

**Comment by bboreham** [comment]
Given a reference to the tracer, it's one line to call `tracer.Inject()` and get the correct formatting.
Perhaps the burden that places on the frontend to extract the trace ID is a consideration.

I agree that `traceparent` should not be abused by using it to send a non-standard value.

**Review by chri2547 — COMMENTED** [review]
Docs approved, thank you!

**Review by mckn — APPROVED** [review]
LGTM!

**Comment by gabor** [comment]
thank you @mckn for the approve 👍 

there are also files changed that are owned by other squads, i just want to make sure everyone's opinion is heard, so i'll ping the squads directly, so they can check the changes:
- @grafana/backend-platform ( `middleware.go`)
- @grafana/grafana-frontend-platform ( `backend_srv.ts`, all the `Inspect*` files)
- @grafana/dashboards-squad ( `runRequest.ts`, `QueryErrorAlert.tsx`)

just want to make sure, are you all OK with the changes? thanks!

File: pkg/middleware/middleware.go [hunk 5]
```diff
@@ -29,12 +30,18 @@ func AddDefaultResponseHeaders(cfg *setting.Cfg) web.Handler {
...
 	return func(c *web.Context) {
 		c.Resp.Before(func(w web.ResponseWriter) { // if response has already been written, skip.
 			if w.Written() {
 				return
 			}
 
+			traceId := tracing.TraceIDFromContext(c.Req.Context(), false)
+			if traceId != "" {
+				w.Header().Set("traceparent", traceId)
+			}
```
Comment (sakjur) [inline pkg/middleware/middleware.go:43]: This doesn't seem to follow the format given by [w3c/trace-context](https://www.w3.org/TR/trace-context-1/#traceparent-header-field-values)?

**Review by sakjur — DISMISSED** [review]
Backend platform: See my [previous comment](https://github.com/grafana/grafana/pull/64903#issuecomment-1479324419) about not using standardized headers unless we also follow the standard. Maybe I'm missing something, but it seems like we're only sending the trace ID right now and not the `<version>-<traceID>-<spanID>-<flags>` style that `traceparent` expects?

File: pkg/middleware/middleware.go (hunk 5 above)
Comment (gabor) [reply pkg/middleware/middleware.go:43]: @sakjur sorry, you are correct, i will fix this (either by giving it a new name or by making it the right format)

Comment (gabor) [reply pkg/middleware/middleware.go]: adjusted it to be "grafana-trace-id" (i can change it even more if needed, no problem)

**Comment by gabor** [comment]
@bboreham @svennergr @sakjur regarding the traceid-header-name discussion... to be able to move forward with this PR that contains a bunch of frontend-code too, i switched to the `grafana-trace-id` header-name for now (this was a smaller diff than the opposite approach.

after this gets merged, i can work on a new PR to change it to something else, if we agree on that, but i did not want to complicate this PR with additional changes (for example, i don't know yet how to get a `tracer` at that place in `middleware.go`, also should we parse both formats of `traceer.Inject()`'s output in javascript or not ? etc.)

basically, i do not consider us changing this header to something else a backward-incompatible breaking change, so i think we can iterate on this separately.

WDYT?

**Comment by sakjur** [comment]
That sounds good to me, we can change it down the line :+1:

**Review by marefr — APPROVED** [review]
SGTM as well

**Review by sakjur — APPROVED** [review]
Tested the insertion of the `Grafana-Trace-Id` header through the middleware (i.e. just the changes to `middleware.go`), and that works well :+1:

**Review by ashharrison90 — APPROVED** [review]
added a question about the classnames, otherwise lgtm 👍

File: packages/grafana-runtime/src/utils/queryResponse.ts
Comment (tskarhed) [inline packages/grafana-runtime/src/utils/queryResponse.ts:1]: Could we add some tests covering these changes?

Comment (gabor) [reply packages/grafana-runtime/src/utils/queryResponse.ts:1]: sure! i originally wanted to get some reviews first before i add the unit tests (i wanted to avoid having to rewrite unit-tests if the architecture is wrong), but now the reviews are rolling in, it's time to add test-coverage 👍

File: packages/grafana-runtime/src/utils/queryResponse.ts [hunk 6]
```diff
@@ -65,6 +65,13 @@ export function toDataQueryResponse(
   queries?: DataQuery[]
 ): DataQueryResponse {
   const rsp: DataQueryResponse = { data: [], state: LoadingState.Done };
+
+  const traceId = 'traceId' in res ? res.traceId : undefined;
```
Comment (axelavargas) [inline packages/grafana-runtime/src/utils/queryResponse.ts:69]: Nit:

```suggestion
  const traceId = rest.traceId ??  undefined;
```

**Review by axelavargas — APPROVED** [review]
Hey Gabor 👋🏾 , code related to `runRequest` good look to me :). I left a small nit 😄 .

**Review by tskarhed — APPROVED** [review]
